Query LaunchServices to determine the host app bundle identifier

LumingYin opened the pull request:
#### 0646de88e5b9e9e657741c81b75e51e353ff35ae
<pre>
Query LaunchServices to determine the host app bundle identifier
<a href="https://bugs.webkit.org/show_bug.cgi?id=255361">https://bugs.webkit.org/show_bug.cgi?id=255361</a>
rdar://107931346

Reviewed by Brady Eidson.

For WebPush on macOS, use LaunchServices instead of the codesigning identifier to determine
the host app identity. If no matching result is found by querying LaunchServices, fallback
to checking the code signing identifier as before.

* Source/WebCore/PAL/pal/spi/cocoa/LaunchServicesSPI.h:
Forward declare LaunchServices SPI.

* Source/WebKit/Configurations/WebKit.xcconfig:
Weak link libbsm so we can extract the audit session ID from the audit token.

* Source/WebKit/Scripts/process-entitlements.sh:
Add the necessary entitlement so we can query LaunchServices with an asn reference.

* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::ClientConnection::hostAppCodeSigningIdentifier):
Call bundleIdentifierFromAuditToken to determine the host bundle identifier on Mac.

(WebPushD::ClientConnection::bundleIdentifierFromAuditToken):
On macOS, prefer using LaunchServices to determine the host app identity, with the code signing
identifier as fallback.

Canonical link: <a href="https://commits.webkit.org/262898@main">https://commits.webkit.org/262898@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a0ab7bd2e1c0e5e7cfd0abde6d575d03c69a653e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2949 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3014 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3111 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4354 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3357 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2916 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3089 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3052 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2584 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2977 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3352 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2639 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4146 "Built successfully") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/2976 "webkitpy-tests (failure)") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/848 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2618 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2479 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2610 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2669 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3891 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3018 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2423 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2651 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2621 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/729 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2638 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2838 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

beidson commented on Source/WebKit/webpushd/PushClientConnection.mm:
```diff
@@ -102,13 +107,38 @@
+    CFStringRef auditTokenKey = CFStringCreateCopy(kCFAllocatorDefault, _kLSAuditTokenKey);
+    CFArrayRef matchingAppsRef = _LSCopyMatchingApplicationsWithItems(sessionID, 1, (CFTypeRef *)(&auditTokenKey), &auditTokenDataRef);
+    CFRelease(auditTokenDataRef);
+    CFRelease(auditTokenKey);
```
WebKit generally prefers smart pointers to avoid manual retain/release counting.
If this wouldn't work, e.g.:
`auto auditTokenDataRef = adoptCF(CFDataCreate(kCFAllocatorDefault, (const UInt8 *)(&audit_token), sizeof(audit_token)));`
Then this would:
`RetainPtr<CFDataRef> auditTokenDataRef = adoptCF(CFDataCreate(kCFAllocatorDefault, (const UInt8 *)(&audit_token), sizeof(audit_token)));`

beidson reviewed (commented):
With smart pointer cleanup, r=me

Commit:
Query LaunchServices to determine the host app bundle identifier
https://bugs.webkit.org/show_bug.cgi?id=255361
rdar://107931346

Reviewed by Brady Eidson.

For WebPush on macOS, use LaunchServices instead of the codesigning identifier to determine
the host app identity. If no matching result is found by querying LaunchServices, fallback
to checking the code signing identifier as before.

* Source/WebCore/PAL/pal/spi/cocoa/LaunchServicesSPI.h:
Forward declare LaunchServices SPI.

* Source/WebKit/Configurations/WebKit.xcconfig:
Weak link libbsm so we can extract the audit session ID from the audit token.

* Source/WebKit/Scripts/process-entitlements.sh:
Add the necessary entitlement so we can query LaunchServices with an asn reference.

* Source/WebKit/webpushd/PushClientConnection.h:
* Source/WebKit/webpushd/PushClientConnection.mm:
(WebPushD::ClientConnection::hostAppCodeSigningIdentifier):
Call bundleIdentifierFromAuditToken to determine the host bundle identifier on Mac.

(WebPushD::ClientConnection::bundleIdentifierFromAuditToken):
On macOS, prefer using LaunchServices to determine the host app identity, with the code signing
identifier as fallback.

Canonical link: https://commits.webkit.org/262898@main

smfr commented on Source/WebKit/webpushd/PushClientConnection.mm:
```diff
@@ -102,13 +107,32 @@
+    CFTypeRef values[] = { _kLSAuditTokenKey };
+    auto matchingAppsRef = adoptCF(_LSCopyMatchingApplicationsWithItems(sessionID, 1, keys, values));
+    if (matchingAppsRef && CFArrayGetCount(matchingAppsRef.get())) {
+        auto asnRef = adoptCF((LSASNRef)CFArrayGetValueAtIndex(matchingAppsRef.get(), 0));
```
This looks like it will over-release. CFArrayGetValueAtIndex() doesn’t return a retained object.

LumingYin replied:
Great catch, I’ve got [a follow-up fix](https://github.com/WebKit/WebKit/pull/12686).
